File: 0001-0500/0016/0016.3-sum-closest.py
#
# @lc app=leetcode id=16 lang=python3
#
# [16] 3Sum Closest
#

# @lc code=start
class Solution:
    def threeSumClosest(self, nums: List[int], target: int) -> int:
        nums.sort()
        closestSum,closestDiff = 0, float("inf")
        length = len(nums)
        endIdx = length-1
        for i in range(length-2):
            currentNum = nums[i]
            left, right = i + 1, endIdx
            while left < right:
                currentSum = currentNum + nums[left] + nums[right]
                currentDiff = abs(currentSum-target)
                if currentDiff == 0:
                    return currentSum
                elif currentDiff < closestDiff:
                    closestDiff = currentDiff
                    closestSum = currentSum
                if currentSum > target:
                    right -= 1
                else:
                    left += 1
                    
        return closestSum
# @lc code=end

File: 0001-0500/0016/0016.3-sum-closest.cs
/*
 * @lc app=leetcode id=16 lang=csharp
 *
 * [16] 3Sum Closest
 */

// @lc code=start
public class Solution {
    public int ThreeSumClosest(int[] nums, int target) {
        Array.Sort(nums);
        int result = nums[0] + nums[1] + nums[2];
        for (int i = 0; i < nums.Length; i++) {
            int left = i + 1;
            int right = nums.Length - 1;
            while (left < right) {
                int sum = nums[i] + nums[left] + nums[right];
                if (Math.Abs(sum - target) < Math.Abs(result - target)) {
                    result = sum;
                }
                if (sum < target) {
                    left++;
                } else {
                    right--;
                }
            }
        }
        return result;  
    }
}
// @lc code=end

